lib/llhttp/parser.js

```diff
@@ -0,0 +1,241 @@
+  }
+}
+
+class HTTPParser {
+  constructor (maxHeaderSize, lenient = false, headersTimeout) {
+    this[kPtr] = inst.exports.llhttp_alloc(constants.TYPE.RESPONSE)
+    this[kBufferSize] = 0
+    this[kBufferPtr] = null
+    this[kCurrentBuffer] = null
+    this[kStatusMessage] = null
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    this[kMaxHeadersSize] = maxHeaderSize
+    this[kHeadersTimeout] = headersTimeout
+
+    if (lenient === true) {
+      inst.exports.llhttp_set_lenient_headers(this[kPtr], 1)
+    }
+  }
+
+  [kMallocBuffer] (size) {
+    if (this[kBufferPtr]) {
+      inst.exports.free(this[kBufferPtr])
+    }
+    this[kBufferSize] = size
+    this[kBufferPtr] = inst.exports.malloc(size)
+  }
+
+  [kGetHeaders] () {
+    const rawHeaders = []
+    for (let c = 0; c < this[kHeadersFields].length; c++) {
+      rawHeaders.push(this[kHeadersFields][c], this[kHeadersValues][c])
+    }
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    return rawHeaders
+  }
+
+  [kOnMessageBegin] () {
+    return 0
+  }
+
+  [kOnHeaders] (rawHeaders) {}
+
+  [kOnHeadersComplete] (versionMajor, versionMinor, rawHeaders, method,
+    url, statusCode, statusMessage, upgrade, shouldKeepAlive) {
+    return 0
+  }
+
+  [kOnBody] (body) {
+    return 0
+  }
+
+  [kOnMessageComplete] () {
+    return 0
+  }
+
+  [kOnExecute] (ret) {}
+
+  close () {
+    inst.exports.llhttp_free(this[kPtr])
+    this[kPtr] = null
+    inst.exports.free(this[kBufferPtr])
+    this[kBufferPtr] = null
+    this[kCurrentBuffer] = Buffer.alloc(0)
+  }
+
+  execute (data) {
+    currentParser = this
+    this[kCurrentBuffer] = data
+    // Be sure the parser buffer can contain `data`
+    if (data.byteLength > this[kBufferSize]) {
+      this[kMallocBuffer](Math.ceil(data.byteLength / 4096) * 4096)
+    }
+    // Instantiate a Unit8 Buffer view of the wasm memory that starts from the parser buffer pointer
+    // and has the size of `data`
+    const u8 = new Uint8Array(inst.exports.memory.buffer, this[kBufferPtr], data.byteLength)
+    // Fill the view with `data`
+    u8.set(data)
+    // Call `execute` on the wasm parser.
+    // We pass the `llhttp_parser` pointer address, the pointer address of buffer view data,
+    // and finally the length of bytes to parse.
+    // The return value is an error code or `constants.ERROR.OK`.
+    // See https://github.com/dnlup/llhttp/blob/undici_wasm/src/native/api.c#L106
+    const err = inst.exports.llhttp_execute(this[kPtr], this[kBufferPtr], data.byteLength)
+    let ret = data.byteLength
+    if (err !== constants.ERROR.OK) {
+      const errorPos = inst.exports.llhttp_get_error_pos(this[kPtr])
+      ret = errorPos - this[kBufferPtr]
+      if (err === constants.ERROR.PAUSED_UPGRADE) {
+        inst.exports.llhttp_resume_after_upgrade(this[kPtr])
+      } else {
+        ret = this[kMakeError](err)
+      }
+    }
+    this[kOnExecute](ret)
+    currentParser = null
+    return ret
```

Comment:
Are we sure that `ret` will return the number of consumed bytes during UPGRADE?

Reply:
I'll check it, thanks

Reply:
I am checking, and I remember implementing https://github.com/nodejs/undici/pull/575/files#diff-a9cec0e255f9220e6501b2976c44d1fdbddf0b706108eb91d07c7ae5f9f58b65R208 mimicking the behavior in the Node parser https://github.com/nodejs/node/blob/master/src/node_http_parser.cc#L743. Before implementing that, the upgrade tests were failing.